### PR TITLE
Support parallel queries / aggregates

### LINKED ISSUE #9: Add combinefunc in aggregates to enable PG10 parallelism
It would be useful to support PG10 parallelism, and also help with https://github.com/citusdata/citus/issues/120

### PR DESCRIPTION
**Pull request by serprex**
Fixes #9 

Also optimize TopnAggState to not have two indirections to hashtable
Also remove unnecessary allocations from PruneHashTable

I believe `join_tests` is failing because parallel aggregation is causing approximations to occur differently

### CODE REVIEW

**Comment by serprex** [comment]
The `join_tests` variance is somewhat worrying. On my own machine it doesn't occur, whereas on travis-ci the variance is non deterministic, thus `join_tests_0.out` represents one output travis gave

**Comment by pykello** [comment]
@serprex are you still working on this? When do you think the tests will pass?

**Comment by serprex** [comment]
@pykello I'm unsure how to proceed

**Comment by marcocitus** [comment]
I think the non-determinism is expected given the non-deterministic ordering of the table, non-deterministic result order from parallel workers, and order-dependent output of TopN, though we should figure out how to get consistent test output.

**Comment by serprex** [comment]
@marcocitus I think I had an alternative expected file not added in git, tests pass on my system & travis now. Or are you wanting tests updated to not have alternative outputs?

File: topn.c [hunk 1]
```diff
@@ -473,6 +473,127 @@ topn_union_trans(PG_FUNCTION_ARGS)
...
+ */
+Datum
+topn_serialize(PG_FUNCTION_ARGS)
+{
+	Size topnArraySize;
+	TopnAggState *topnTrans = (TopnAggState *) PG_GETARG_POINTER(0);
+	HASH_SEQ_STATUS status;
+	FrequentTopnItem *currentTask = NULL;
+	bytea *ret;
+	char *bpPtr; /* Cursor for writing into ret */
```
Comment (marcocitus) [inline topn.c:491]: variable name could be improved

File: topn.c [hunk 2]
```diff
@@ -473,6 +473,127 @@ topn_union_trans(PG_FUNCTION_ARGS)
...
+	topnArraySize = sizeof(FrequentTopnItem) * hashTableSize;
+	ret = palloc(VARHDRSZ + topnArraySize);
+	SET_VARSIZE(ret, VARHDRSZ + topnArraySize);
+	bpPtr = (void *) VARDATA(ret);
+
+	hash_seq_init(&status, topnHashtable(topnTrans));
+
+	while ((currentTask = (FrequentTopnItem *) hash_seq_search(&status)) != NULL)
+	{
+		memcpy(bpPtr, currentTask, sizeof(FrequentTopnItem));
```
Comment (marcocitus) [inline topn.c:511]: Minor note: It strikes me that a lot of these serialize functions might have more subtle over-the-network issues like the potential for different character encodings across servers. We can probably ignore that here.

File: Makefile [hunk 3]
```diff
@@ -2,19 +2,24 @@
...
 REGRESS = add_agg union_agg char_tests null_tests add_union_tests copy_data customer_reviews_query join_tests
 
+
 # be explicit about the default target
 all:
 
+SQLPP ?= cpp -undef -w -P -imacros $(shell $(PG_CONFIG) --includedir-server)/pg_config.h
+
+%.sql: update/%.sql
+	$(SQLPP) $^ > $@
```
Comment (marcocitus) [inline Makefile:19]: we should make add the generated SQL upgrade scripts to `make clean`

**Review by marcocitus — APPROVED** [review]

**Comment by marcocitus** [comment]
>  I think I had an alternative expected file not added in git, tests pass on my system & travis now. Or are you wanting tests updated to not have alternative outputs?

alternative output seems ok

**Pull request merged**